File: fabric-network/chaincode/olivechain-domain/workflow.go
// SPDX-License-Identifier: Apache-2.0
package main

import (
	"fmt"
	"sort"
	"strings"

	"github.com/hyperledger/fabric-contract-api-go/v2/contractapi"
)

const (
	WorkflowProduct  = "product"
	WorkflowCircular = "circular"
	WorkflowUnknown  = "unknown"
)

var productWorkflow = []string{
	EventCultivation,
	EventHarvest,
	EventCollectionTransport,
	EventMilling,
	EventLabVerification,
	EventBottling,
	EventDistributionRetail,
}

var circularWorkflow = []string{
	EventResidueGeneration,
	EventResidueCustody,
	EventValorization,
	EventUsefulOutput,
	EventReturnOrSale,
	EventEnvAccounting,
}

var productStageLabels = map[string]string{
	EventCultivation:         "Farming recorded",
	EventHarvest:             "Harvest recorded",
	EventCollectionTransport: "Olives transported",
	EventMilling:             "Olives milled",
	EventLabVerification:     "Quality verified",
	EventBottling:            "Oil bottled",
	EventDistributionRetail:  "Delivered to seller",
}

var circularStageLabels = map[string]string{
	EventResidueGeneration: "Residue generated",
	EventResidueCustody:    "Residue in custody transfer",
	EventValorization:      "Residue valorized",
	EventUsefulOutput:      "Useful output recorded",
	EventReturnOrSale:      "Output returned or sold",
	EventEnvAccounting:     "Environmental accounting complete",
}

func workflowForEvent(eventType string) (string, []string, bool) {
	if containsString(productWorkflow, eventType) {
		return WorkflowProduct, productWorkflow, true
	}
	if containsString(circularWorkflow, eventType) {
		return WorkflowCircular, circularWorkflow, true
	}
	return WorkflowUnknown, nil, false
}

func workflowIndex(stages []string, eventType string) int {
	for i, stage := range stages {
		if stage == eventType {
			return i
		}
	}
	return -1
}

func stageLabel(workflow, stage string) string {
	if stage == "" || stage == "not_started" {
		return "Not started"
	}
	if workflow == WorkflowProduct {
		if label := productStageLabels[stage]; label != "" {
			return label
		}
	}
	if workflow == WorkflowCircular {
		if label := circularStageLabels[stage]; label != "" {
			return label
		}
	}
	return strings.ReplaceAll(stage, "_", " ")
}

func subjectWorkflowEvents(ctx contractapi.TransactionContextInterface, subjectID string) ([]*Event, error) {
	events, err := listEventsFromIndex(ctx, subjectObjectType, []string{subjectID}, false)
	if err != nil {
		return nil, err
	}
	relevant := make([]*Event, 0, len(events))
	for _, event := range events {
		if _, _, ok := workflowForEvent(event.EventType); ok {
			relevant = append(relevant, event)
		}
	}
	return relevant, nil
}

func analyzeWorkflow(subjectID string, events []*Event) (*WorkflowStatus, error) {
	status := &WorkflowStatus{
		DocType:           "workflow_status",
		SubjectID:         subjectID,
		Workflow:          WorkflowUnknown,
		Exists:            false,
		CurrentStep:       "not_started",
		CurrentLabel:      "Not started",
		CurrentStepNumber: 0,
		TotalSteps:        0,
		ProgressPct:       0,
		NextAllowed:       []string{},
		Complete:          false,
		Valid:             true,
		Message:           "No workflow event has been recorded for this subject.",
		LastEventID:       "",
		UpdatedAt:         "",
		OriginBatch:       "",
	}
	if len(events) == 0 {
		return status, nil
	}

	workflow := WorkflowUnknown
	var stages []string
	for _, event := range events {
		candidate, candidateStages, ok := workflowForEvent(event.EventType)
		if !ok {
			continue
		}
		if workflow == WorkflowUnknown {
			workflow = candidate
			stages = candidateStages
		} else if workflow != candidate {
			status.Valid = false
			status.Message = fmt.Sprintf("subject %s mixes product and circular workflow events", subjectID)
			return status, nil
		}
	}
	if workflow == WorkflowUnknown {
		return status, nil
	}

	status.Workflow = workflow
	status.Exists = true
	status.TotalSteps = len(stages)
	current := -1
	acceptedCustody := false
	seen := map[string]int{}

	for _, event := range events {
		candidate, _, ok := workflowForEvent(event.EventType)
		if !ok || candidate != workflow {
			continue
		}
		idx := workflowIndex(stages, event.EventType)
		if idx < 0 {
			continue
		}
		seen[event.EventType]++

		switch workflow {
		case WorkflowProduct:
			if idx != current+1 {
				status.Valid = false
				expected := stages[0]
				if current+1 < len(stages) {
					expected = stages[current+1]
				}
				status.Message = fmt.Sprintf(
					"existing history is out of order: found %s while expecting %s",
					event.EventType, expected,
				)
				return status, nil
			}
			current = idx
		case WorkflowCircular:
			if event.EventType == EventResidueCustody && current == 1 {
				if asBool(event.Payload["accepted"]) {
					acceptedCustody = true
				}
				status.LastEventID = event.EventID
				status.UpdatedAt = event.TimestampRFC
				continue
			}
			if event.EventType == EventUsefulOutput && current == 3 {
				status.LastEventID = event.EventID
				status.UpdatedAt = event.TimestampRFC
				continue
			}
			if event.EventType == EventReturnOrSale && current == 4 {
				status.LastEventID = event.EventID
				status.UpdatedAt = event.TimestampRFC
				continue
			}
			if idx != current+1 {
				status.Valid = false
				expected := stages[0]
				if current+1 < len(stages) {
					expected = stages[current+1]
				}
				status.Message = fmt.Sprintf(
					"existing circular history is out of order: found %s while expecting %s",
					event.EventType, expected,
				)
				return status, nil
			}
			current = idx
			if event.EventType == EventResidueCustody && asBool(event.Payload["accepted"]) {
				acceptedCustody = true
			}
			if event.EventType == EventResidueGeneration {
				status.OriginBatch = asString(event.Payload["origin_batch"])
			}
		}
		status.LastEventID = event.EventID
		status.UpdatedAt = event.TimestampRFC
	}

	if current >= 0 {
		status.CurrentStep = stages[current]
		status.CurrentLabel = stageLabel(workflow, status.CurrentStep)
		status.CurrentStepNumber = current + 1
		status.ProgressPct = round2(float64(current+1) * 100 / float64(len(stages)))
	}

	if current == len(stages)-1 {
		status.Complete = true
		status.Message = "Workflow complete."
		if workflow == WorkflowProduct {
			status.NextAllowed = []string{EventConsumerVerification}
		}
		return status, nil
	}

	if workflow == WorkflowCircular {
		switch current {
		case 1:
			if acceptedCustody {
				status.NextAllowed = []string{EventValorization}
				status.Message = "Residue accepted. Valorization is the next required step."
			} else {
				status.NextAllowed = []string{EventResidueCustody}
				status.Message = "Waiting for an accepted residue custody record before valorization."
			}
			return status, nil
		case 3:
			status.NextAllowed = []string{EventUsefulOutput, EventReturnOrSale}
			status.Message = "Additional useful outputs may be recorded, or the workflow may continue to return or sale."
			return status, nil
		case 4:
			status.NextAllowed = []string{EventReturnOrSale, EventEnvAccounting}
			status.Message = "Additional destinations may be recorded, or environmental accounting may complete the workflow."
			return status, nil
		}
	}

	status.NextAllowed = []string{stages[current+1]}
	status.Message = fmt.Sprintf("Next required step: %s.", stageLabel(workflow, stages[current+1]))
	return status, nil
}

func validateWorkflowTransition(ctx contractapi.TransactionContextInterface, eventType, subjectID string, payload map[string]interface{}) error {
	if eventType == EventCorrection || eventType == EventRewardIssued || eventType == EventReviewFlag || eventType == EventReviewResolution {
		return nil
	}

	if eventType == EventConsumerVerification {
		events, err := subjectWorkflowEvents(ctx, subjectID)
		if err != nil {
			return err
		}
		status, err := analyzeWorkflow(subjectID, events)
		if err != nil {
			return err
		}
		if !status.Exists || status.Workflow != WorkflowProduct || !status.Complete {
			return fmt.Errorf("consumer verification is allowed only after batch %s reaches distribution_retail", subjectID)
		}
		return nil
	}

	workflow, stages, managed := workflowForEvent(eventType)
	if !managed {
		return nil
	}

	if workflow == WorkflowCircular {
		if eventType != EventResidueGeneration {
			residueID := strings.TrimSpace(asString(payload["residue_id"]))
			if residueID == "" || residueID != strings.TrimSpace(subjectID) {
				return fmt.Errorf("%s requires payload residue_id to match subject_id %s", eventType, subjectID)
			}
		}
		if eventType == EventResidueGeneration {
			originBatch := strings.TrimSpace(asString(payload["origin_batch"]))
			if originBatch == "" {
				return fmt.Errorf("residue_generation requires origin_batch")
			}
			batchEvents, err := subjectWorkflowEvents(ctx, originBatch)
			if err != nil {
				return err
			}
			batchStatus, err := analyzeWorkflow(originBatch, batchEvents)
			if err != nil {
				return err
			}
			millingIndex := workflowIndex(productWorkflow, EventMilling)
			if !batchStatus.Exists || batchStatus.Workflow != WorkflowProduct || batchStatus.CurrentStepNumber < millingIndex+1 {
				return fmt.Errorf("residue %s cannot be generated before origin batch %s reaches milling", subjectID, originBatch)
			}
		}
	}

	events, err := subjectWorkflowEvents(ctx, subjectID)
	if err != nil {
		return err
	}
	status, err := analyzeWorkflow(subjectID, events)
	if err != nil {
		return err
	}
	if !status.Valid {
		return fmt.Errorf("cannot advance %s: %s", subjectID, status.Message)
	}

	if !status.Exists {
		if eventType != stages[0] {
			return fmt.Errorf("%s has not started; first required event is %s, not %s", subjectID, stages[0], eventType)
		}
		return nil
	}
	if status.Workflow != workflow {
		return fmt.Errorf("subject %s belongs to %s workflow, not %s", subjectID, status.Workflow, workflow)
	}
	if status.Complete {
		return fmt.Errorf("%s workflow for %s is already complete; %s is not allowed", workflow, subjectID, eventType)
	}
	if containsString(status.NextAllowed, eventType) {
		return nil
	}
	if status.CurrentStep == eventType {
		return fmt.Errorf("%s already has step %s; next allowed event is %s", subjectID, eventType, strings.Join(status.NextAllowed, " or "))
	}
	return fmt.Errorf(
		"workflow order violation for %s: current step is %s; next allowed event is %s; cannot submit %s",
		subjectID, status.CurrentStep, strings.Join(status.NextAllowed, " or "), eventType,
	)
}

func (c *OliveChainContract) GetSubjectStatus(ctx contractapi.TransactionContextInterface, subjectID string) (*WorkflowStatus, error) {
	events, err := subjectWorkflowEvents(ctx, subjectID)
	if err != nil {
		return nil, err
	}
	return analyzeWorkflow(subjectID, events)
}

func (c *OliveChainContract) GetWorkflowStatuses(ctx contractapi.TransactionContextInterface) ([]WorkflowStatus, error) {
	events, err := listEventsFromIndex(ctx, eventObjectType, nil, false)
	if err != nil {
		return nil, err
	}
	bySubject := map[string][]*Event{}
	for _, event := range events {
		if _, _, ok := workflowForEvent(event.EventType); !ok {
			continue
		}
		bySubject[event.SubjectID] = append(bySubject[event.SubjectID], event)
	}
	subjects := make([]string, 0, len(bySubject))
	for subjectID := range bySubject {
		subjects = append(subjects, subjectID)
	}
	sort.Strings(subjects)
	statuses := make([]WorkflowStatus, 0, len(subjects))
	for _, subjectID := range subjects {
		status, err := analyzeWorkflow(subjectID, bySubject[subjectID])
		if err != nil {
			return nil, err
		}
		statuses = append(statuses, *status)
	}
	return statuses, nil
}

func (c *OliveChainContract) GetWorkflowDefinitions(ctx contractapi.TransactionContextInterface) ([]WorkflowDefinition, error) {
	return []WorkflowDefinition{
		{Name: WorkflowProduct, Steps: append([]string{}, productWorkflow...), Repeatable: []string{EventConsumerVerification}},
		{Name: WorkflowCircular, Steps: append([]string{}, circularWorkflow...), Repeatable: []string{EventResidueCustody, EventUsefulOutput, EventReturnOrSale}},
	}, nil
}

File: olivechain/workflow.py
"""Ordered OliveChain product and circular workflow helpers.

Fabric chaincode is authoritative for write enforcement. This module mirrors
its status model for the FastAPI UI and for the optional local backend.
"""
from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Iterable

PRODUCT_STEPS = [
    "cultivation",
    "harvest",
    "collection_transport",
    "milling",
    "lab_verification",
    "bottling",
    "distribution_retail",
]

CIRCULAR_STEPS = [
    "residue_generation",
    "residue_custody",
    "valorization",
    "useful_output",
    "return_or_sale",
    "env_accounting",
]

PRODUCT_LABELS = {
    "cultivation": "Farming recorded",
    "harvest": "Harvest recorded",
    "collection_transport": "Olives transported",
    "milling": "Olives milled",
    "lab_verification": "Quality verified",
    "bottling": "Oil bottled",
    "distribution_retail": "Delivered to seller",
}

CIRCULAR_LABELS = {
    "residue_generation": "Residue generated",
    "residue_custody": "Residue in custody transfer",
    "valorization": "Residue valorized",
    "useful_output": "Useful output recorded",
    "return_or_sale": "Output returned or sold",
    "env_accounting": "Environmental accounting complete",
}


@dataclass
class WorkflowStatus:
    subject_id: str
    workflow: str = "unknown"
    exists: bool = False
    current_step: str = "not_started"
    current_label: str = "Not started"
    current_step_number: int = 0
    total_steps: int = 0
    progress_pct: float = 0.0
    next_allowed: list[str] | None = None
    complete: bool = False
    valid: bool = True
    message: str = "No workflow event has been recorded for this subject."
    last_event_id: str = ""
    updated_at: str = ""
    origin_batch: str = ""

    def to_dict(self) -> dict:
        data = asdict(self)
        data["next_allowed"] = list(self.next_allowed or [])
        return data


def _event_type(event) -> str:
    return getattr(event, "event_type", "")


def _event_payload(event) -> dict:
    value = getattr(event, "payload", {})
    return value if isinstance(value, dict) else {}


def _event_timestamp(event) -> str:
    timestamp = getattr(event, "timestamp", 0)
    return str(timestamp or "")


def _workflow_for_type(event_type: str):
    if event_type in PRODUCT_STEPS:
        return "product", PRODUCT_STEPS
    if event_type in CIRCULAR_STEPS:
        return "circular", CIRCULAR_STEPS
    return None, None


def derive_status(subject_id: str, events: Iterable) -> WorkflowStatus:
    relevant = [event for event in events if _workflow_for_type(_event_type(event))[0]]
    status = WorkflowStatus(subject_id=subject_id, next_allowed=[])
    if not relevant:
        return status

    workflows = {_workflow_for_type(_event_type(event))[0] for event in relevant}
    if len(workflows) != 1:
        status.exists = True
        status.valid = False
        status.message = f"Subject {subject_id} mixes product and circular workflow events."
        return status

    workflow = next(iter(workflows))
    steps = PRODUCT_STEPS if workflow == "product" else CIRCULAR_STEPS
    labels = PRODUCT_LABELS if workflow == "product" else CIRCULAR_LABELS
    status.workflow = workflow
    status.exists = True
    status.total_steps = len(steps)
    current = -1
    accepted_custody = False

    for event in relevant:
        event_type = _event_type(event)
        idx = steps.index(event_type)
        payload = _event_payload(event)

        if workflow == "product":
            if idx != current + 1:
                expected = steps[min(current + 1, len(steps) - 1)]
                status.valid = False
                status.message = (
                    f"Existing history is out of order: found {event_type} "
                    f"while expecting {expected}."
                )
                return status
            current = idx
        else:
            if event_type == "residue_custody" and current == 1:
                accepted_custody = accepted_custody or bool(payload.get("accepted"))
                status.last_event_id = getattr(event, "event_id", "")
                status.updated_at = _event_timestamp(event)
                continue
            if event_type == "useful_output" and current == 3:
                status.last_event_id = getattr(event, "event_id", "")
                status.updated_at = _event_timestamp(event)
                continue
            if event_type == "return_or_sale" and current == 4:
                status.last_event_id = getattr(event, "event_id", "")
                status.updated_at = _event_timestamp(event)
                continue
            if idx != current + 1:
                expected = steps[min(current + 1, len(steps) - 1)]
                status.valid = False
                status.message = (
                    f"Existing circular history is out of order: found {event_type} "
                    f"while expecting {expected}."
                )
                return status
            current = idx
            if event_type == "residue_custody":
                accepted_custody = bool(payload.get("accepted"))
            if event_type == "residue_generation":
                status.origin_batch = str(payload.get("origin_batch", ""))

        status.last_event_id = getattr(event, "event_id", "")
        status.updated_at = _event_timestamp(event)

    if current >= 0:
        status.current_step = steps[current]
        status.current_label = labels[steps[current]]
        status.current_step_number = current + 1
        status.progress_pct = round((current + 1) * 100 / len(steps), 2)

    if current == len(steps) - 1:
        status.complete = True
        status.message = "Workflow complete."
        status.next_allowed = ["consumer_verification"] if workflow == "product" else []
        return status

    if workflow == "circular":
        if current == 1:
            if accepted_custody:
                status.next_allowed = ["valorization"]
                status.message = "Residue accepted. Valorization is the next required step."
            else:
                status.next_allowed = ["residue_custody"]
                status.message = "Waiting for an accepted custody record before valorization."
            return status
        if current == 3:
            status.next_allowed = ["useful_output", "return_or_sale"]
            status.message = "Record another useful output or continue to return or sale."
            return status
        if current == 4:
            status.next_allowed = ["return_or_sale", "env_accounting"]
            status.message = "Record another destination or complete environmental accounting."
            return status

    status.next_allowed = [steps[current + 1]]
    status.message = f"Next required step: {labels[steps[current + 1]]}."
    return status


def statuses_from_events(events: Iterable) -> list[dict]:
    by_subject: dict[str, list] = {}
    for event in events:
        if not _workflow_for_type(_event_type(event))[0]:
            continue
        by_subject.setdefault(getattr(event, "subject_id", ""), []).append(event)
    return [derive_status(subject, by_subject[subject]).to_dict() for subject in sorted(by_subject)]


def transition_allowed(status: dict, event_type: str) -> tuple[bool, str]:
    if event_type in {"correction", "reward_issued", "review_flag", "review_resolution"}:
        return True, ""
    if event_type == "consumer_verification":
        if status.get("workflow") == "product" and status.get("complete"):
            return True, ""
        return False, "Consumer verification is available only after distribution."
    if not status.get("exists"):
        first = "cultivation" if event_type in PRODUCT_STEPS else "residue_generation"
        if event_type == first:
            return True, ""
        return False, f"The workflow has not started. First required event: {first}."
    if event_type in (status.get("next_allowed") or []):
        return True, ""
    next_text = " or ".join(status.get("next_allowed") or []) or "none"
    return False, (
        f"Current step is {status.get('current_step')}; "
        f"next allowed event is {next_text}; cannot submit {event_type}."
    )
